### DISCUSSION TITLE
help needed with types

### DISCUSSION THREAD

**Discussion by danish-mehmood** [MathisBullinger/froebel#12] [Q&A] [answered]
hey everyone i am reading the froebel code base and learning so much 
but there are some types which are i can't understand can anyone tell what the following types are doing 
-> one is 'λ' type (can not understand what its doing)
-> and the second is 'Widen' type 
thanks

**Comment by MathisBullinger** [comment, marked as answer]
Hi Danish,

the `λ` type is just a shortcut for a function type. So the type `λ` is equivalent to the type `(...args: any[]) => any` or in other words a function that takes any number of arguments and returns anything. The type can also take two optional parameters: The first one is a list of the arguments the function accepts and the second is the function's return type. I.e. `λ<[a: number, b?: string], number>` is equivalent to writing `(a: number, b?: string) => number`. The same type is also exported under the name `fun` from [types.ts](https://github.com/MathisBullinger/froebel/blob/main/types.ts#L2).

The `Widen` type converts string unions to a generic `string` type and number unions to a generic `number`. So `Widen<'a' | 'b'>` returns the type `string` instead of the union `'a' | 'b'` and the same is true for numbers. If the parameter is not a string or number then `Widen` just returns that parameter as is. So in a sense, it "widens" the stricter union types into their more generic equivalents.

Most of the types in `types.ts` are only meant to be used internally by the utilities in this library, hence the lack of documentation for most of them. But I have been thinking about cleaning them up a bit, adding documentation for every type, and moving them into a type-only library that froebel would depend on at some point.

**Reply by danish-mehmood** [reply]
hey mathis thanks for answering , i even emailed you , but you were so swift to reply thanks alot 
and the idea of documenting the internal types as well is awesome 
i am reading all the projects you have on github and learning so much
